great_expectations_provider/operators/great_expectations.py

```diff
@@ -251,48 +252,97 @@ def make_connection_string(self) -> str:
+
+                private_key_file = conn.extra_dejson.get(
+                    "extra__snowflake__private_key_file"
+                ) or conn.extra_dejson.get("private_key_file")
+
+                if private_key_file:
+                    private_key_pem = Path(private_key_file).read_bytes()
+
+                    passphrase = None
+                    if conn.password:
+                        passphrase = conn.password.strip().encode()
+
+                    p_key = serialization.load_pem_private_key(
+                        private_key_pem, password=passphrase, backend=default_backend()
+                    )
+
+                    pkb = p_key.private_bytes(
+                        encoding=serialization.Encoding.DER,
+                        format=serialization.PrivateFormat.PKCS8,
+                        encryption_algorithm=serialization.NoEncryption(),
+                    )
+                    return {
+                        # Unfortunately GE uses deepcopy when instantiating the SqlAlchemyExecutionEngine
+                        # which uses pickle and SAEngine is not pickleable.
+                        # "engine": engine,
+                        "url": url,
+                        "connect_args": {
+                            "private_key": pkb,
+                        },
+                    }
 
-            uri_string = f"snowflake://{self.conn.login}:{self.conn.password}@{snowflake_account}.{snowflake_region}/{snowflake_database}/{self.schema}?warehouse={snowflake_warehouse}&role={snowflake_role}"  # noqa
+                return {"url": url}
+
+            except ImportError:
+                self.log.warning(
+                    (
+                        "Snowflake provider package not available, "
+                        "attempt to manually build connection. "
+                        "Key-based auth is not supported."
+                    )
+                )
+
+                snowflake_account = (
+                    self.conn.extra_dejson.get("account") or self.conn.extra_dejson["extra__snowflake__account"]
+                )
+                snowflake_region = (
+                    self.conn.extra_dejson.get("region") or self.conn.extra_dejson["extra__snowflake__region"]
+                )
+                snowflake_database = (
+                    self.conn.extra_dejson.get("database") or self.conn.extra_dejson["extra__snowflake__database"]
+                )
+                snowflake_warehouse = (
+                    self.conn.extra_dejson.get("warehouse") or self.conn.extra_dejson["extra__snowflake__warehouse"]
+                )
+                snowflake_role = self.conn.extra_dejson.get("role") or self.conn.extra_dejson["extra__snowflake__role"]
+
+                uri_string = f"snowflake://{self.conn.login}:{self.conn.password}@{snowflake_account}.{snowflake_region}/{snowflake_database}/{self.schema}?warehouse={snowflake_warehouse}&role={snowflake_role}"  # noqa
```

Comment:
Is this path only going to be taken if an ImportError occurs? I think the check for a private key file needs to happen first, then the `try` for importing packages makes sense.

Reply:
Isn't it better to use the `SnowflakeHook` if it is available? This isn't just covering the private key scenario. There are other cases that the hook handles gracefully. For example, if the connection has no role defined.

Reply:
So for this case I do see that the url is different than the connection uri, so it does make sense to use SnowflakeHook. This is getting complicated enough that I think the Snowflake portion can be moved out to it's own function, but I don't think the logic has to change.

Reply:
I moved this into its own function.